File: app/frontend/src/lib/api/habits.ts
/**
 * Habits API
 *
 * API client methods for habit tracking.
 * All calls go through the backend at api.ecent.online.
 *
 * PARITY-026: Habits routes
 * REFACTOR: Uses shared client (January 2026)
 */

import { apiGet, apiPost } from './client';

// ============================================
// Types
// ============================================

export interface Habit {
  id: string;
  name: string;
  description: string | null;
  frequency: 'daily' | 'weekly' | 'custom';
  target_count: number;
  icon: string | null;
  color: string | null;
  is_active: boolean;
  current_streak: number;
  longest_streak: number;
  last_completed_at: string | null;
  completed_today: boolean;
  sort_order: number;
}

export interface HabitsList {
  habits: Habit[];
}

export interface CreateHabitRequest {
  name: string;
  description?: string;
  frequency?: 'daily' | 'weekly' | 'custom';
  target_count?: number;
  icon?: string;
  color?: string;
}

export interface CompleteHabitResult {
  habit: Habit;
  new_streak: number;
  xp_awarded: number;
  streak_bonus: boolean;
}

// ============================================
// API Methods
// ============================================

/**
 * List all active habits with today's completion status
 * GET /api/habits
 */
export async function listHabits(): Promise<HabitsList> {
  const response = await apiGet<{ data: HabitsList }>('/api/habits');
  return response.data;
}

/**
 * Create a new habit
 * POST /api/habits
 */
export async function createHabit(req: CreateHabitRequest): Promise<Habit> {
  const response = await apiPost<{ data: Habit }>('/api/habits', req);
  return response.data;
}

/**
 * Complete a habit for today
 * POST /api/habits/:id/complete
 */
export async function completeHabit(habitId: string, notes?: string): Promise<CompleteHabitResult> {
  const response = await apiPost<{ data: CompleteHabitResult }>(
    `/api/habits/${habitId}/complete`,
    notes ? { notes } : undefined
  );
  return response.data;
}

// ============================================
// React Query Keys
// ============================================

export const habitsKeys = {
  all: ['habits'] as const,
  list: () => [...habitsKeys.all, 'list'] as const,
};

File: app/backend/crates/api/src/routes/habits.rs
//! Habits routes
//!
//! Routes for habit tracking.

use std::sync::Arc;

use axum::{
    extract::{Extension, Path, State},
    routing::{get, post},
    Json, Router,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::db::habits_goals_models::*;
use crate::db::habits_goals_repos::HabitsRepo;
use crate::db::models::User;
use crate::error::AppError;
use crate::state::AppState;

/// Create habits routes
pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(list_habits).post(create_habit))
        .route("/{id}/complete", post(complete_habit))
}

// ============================================================================
// RESPONSE WRAPPERS
// ============================================================================

#[derive(Serialize)]
struct HabitResponseWrapper {
    habit: HabitResponse,
}

#[derive(Serialize)]
struct HabitsListWrapper {
    habits: Vec<HabitResponse>,
}

#[derive(Serialize)]
struct CompleteResultWrapper {
    result: CompleteHabitResult,
}

// ============================================================================
// HANDLERS
// ============================================================================

/// GET /habits
/// List active habits with today's completion status
async fn list_habits(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<User>,
) -> Result<Json<HabitsListWrapper>, AppError> {
    let result = HabitsRepo::list_active(&state.db, user.id).await?;

    Ok(Json(HabitsListWrapper { habits: result.habits }))
}

/// POST /habits
/// Create a new habit
async fn create_habit(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<User>,
    Json(req): Json<CreateHabitRequest>,
) -> Result<Json<HabitResponseWrapper>, AppError> {
    let habit = HabitsRepo::create(&state.db, user.id, &req).await?;

    Ok(Json(HabitResponseWrapper {
        habit: HabitResponse {
            id: habit.id,
            name: habit.name,
            description: habit.description,
            frequency: habit.frequency,
            target_count: habit.target_count,
            icon: habit.icon,
            color: habit.color,
            is_active: habit.is_active,
            current_streak: habit.current_streak,
            longest_streak: habit.longest_streak,
            last_completed_at: habit.last_completed_at,
            completed_today: false,
            sort_order: habit.sort_order,
        },
    }))
}

#[derive(Debug, Deserialize)]
struct CompleteHabitBody {
    notes: Option<String>,
}

/// POST /habits/:id/complete
/// Complete a habit for today
async fn complete_habit(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<User>,
    Path(id): Path<Uuid>,
    body: Option<Json<CompleteHabitBody>>,
) -> Result<Json<CompleteResultWrapper>, AppError> {
    let notes = body.and_then(|b| b.notes.clone());
    let result = HabitsRepo::complete_habit(&state.db, id, user.id, notes.as_deref()).await?;

    Ok(Json(CompleteResultWrapper { result }))
}
